### ISSUE TITLE
Bug: Outpost not deleting on MySQL databases

### ISSUE THREAD

**Issue by Portanna** [LlmDl/Towny#2622]
After unclaiming an outpost, the outpost is still present in the database and people can still teleport to its spawn coordinates with /t outpost

### What steps will reproduce the problem?
1. Create an outpost and claim some chunks surrounding its main claim
2. Unclaim that outpost
3. Save-all, shutdown the server, watch in the MySQL database and the outpost spawn is still there, even after rebooting the players can still teleport to the outpost

### What is your Towny version?
I noticed this bug when using 0.91.4.0 (release build) and it's still present in the 0.91.4.7 dev build.

### What is your spigot/craftbukkit version?
Spigot 1.11.2, but the bug was present also when using 1.9.4, 1.10.2 and 1.8.8

Tested with default config (just with the mySQL database set-up), on a clean brand new server using just Towny and its dependencies (Vault for permissions and EssentialsX for economy)

**Comment by LlmDl** [comment]
Does this happen when they're unclaiming their only outpost? Does it happen if they're unclaiming their second or third outpost?

**Comment by Portanna** [comment]
Seems to happen in any case, they just don't get deleted at all from the database

**Comment by LlmDl** [comment]
Can you test it for me?

**Comment by Portanna** [comment]
Yeah, anyways I've been using towny since 2015. And since 2015 there is a town which has old outpost still set on worlds that are not even enabled in towny anymore. And this town has many outposts and has deleted some in its history.

**Comment by LlmDl** [comment]
Going over the code I think I can see how trace outpostspawns would occur, but only if there were only one outpost before it was removed. If I'm right, a town with two outposts that removes one, will work properly.

**Comment by LlmDl** [comment]
I've got a fix for what I think is the issue, in 0.91.4.8, I'd still like you to test whether or not my above assumption is correct.

**Comment by LlmDl** [comment]
http://towny.palmergames.com/?p=1001 Grab this dev and let me know if you can reproduce the issue you're seeing.

**Comment by Portanna** [comment]
Testing in progress...

**Comment by LlmDl** [comment]
Should I take it from your silence that it works fine now?

**Comment by Portanna** [comment]
Testing with the  0.91.4.7 dev build: The outposts spawns are not deleted from the database in any case.

Testing with the dev build you provided: Can't startup server as columns for the "outlaw" are not created.

**Comment by LlmDl** [comment]
Please give it another go using 0.91.4.9 to test, it is available on the website.

**Comment by Portanna** [comment]
Now the new outposts delete correctly but the old leftovers are still there.

**Comment by LlmDl** [comment]
Alright, that's a start, I'll see if I can code up something that would remove invalid outposts (ones that aren't inside outposts owned by the town anymore.)

**Comment by Portanna** [comment]
Well great

**Comment by LlmDl** [comment]
Please re-test using 0.91.4.12, released today. We're going to put a close eye on this issue but we need up to date feedback.

**Comment by Portanna** [comment]
Testing in progress...

**Comment by Portanna** [comment]
After updating to 0.91.4.12, Towny could not recognize any of the current registered nations, resulting in a SAFE MODE, thus rendering this dev build broken and unusable.

**Comment by LlmDl** [comment]
@Portanna can you file a new ticket for that?

**Comment by Articdive** [comment]
@Portanna won't be necessary my fault for something reallllly stupid xD. I'd say 0.91.4.12 should be updated with a quick hotfix no idea if it will happen or not.

**Comment by Portanna** [comment]
Ok let me know as soon as you post the hotfix so that I can test further.

**Comment by LlmDl** [comment]
@Portanna https://www.dropbox.com/s/q6t4kp5hr1u1wkp/Towny.jar?dl=0 Try that please.

**Comment by Portanna** [comment]
Testing in progress...

**Comment by Portanna** [comment]
The build you provided seems to work fine, further testing is in progress but at least the database was correctly loaded this time.

**Comment by LlmDl** [comment]
Articdive did some more looking and there's a pending PR that should fully fix the issue.

Sadly we can only clean up some of the already bugged outpostspawns which are in older DBs. There shouldn't be any newly bugged outpostspawns in the next dev build. I'd still love to hear what you get from that build.

**Comment by Portanna** [comment]
Isn't there any way to speed up the loading time of the database? (not complaining, just curious)

Regarding the testing, I'll wait for the players to give me some feedback tomorrow, if I find out something's not right, I'll write down here as usual.

**Comment by LlmDl** [comment]
The only real way to make the database load faster is to have a smaller database. If you have teh revert-on-unclaim feature turned on, you can consider turning it off in each world and deleting the plot-block-data folder. That will speed things up quite a bit.

**Comment by Portanna** [comment]
Logs after loading the existing database with latest version: https://pastebin.com/B2894uEm

**Comment by LlmDl** [comment]
Thanks, there'll be a new build we'll need tested in the next day or so.

**Comment by Articdive** [comment]
The Pull request for .14 Should:
Allow outposts to load correctly
Remove glitched town outposts /t outpost (number)
As for the nulled TownBlocks (https://pastebin.com/B2894uEm) that's something that will need further investigation.

**Comment by LlmDl** [comment]
Ok new build for you to test @Portanna https://www.dropbox.com/s/q6t4kp5hr1u1wkp/Towny.jar?dl=0

**Comment by Portanna** [comment]
Ok, I'll test this in a few hours

**Comment by Portanna** [comment]
Works fine, but still throws this when enabling: https://pastebin.com/Ly9UQwk4

**Comment by LlmDl** [comment]
Hmm, the first one is about 21 townblocks that are still in the DB, but not registered to a town. It could be helpful to see what one of them has for an entry in the DB, if you can look that up for us.

The second startup, maybe you could toggle on debug mode in the config and do a startup and it will give us more information. 

Aside from that, you've tested claiming multiple outposts, unclaiming them? How many did you claim at a time? The original bug would only show up when the last outpost was unclaimed.

**Comment by Portanna** [comment]
I'm testing the outpost claim/unclaim, regarding the townblocks that are still in the DB, their town is no longer registered and also the world is no longer claimable and does not have any townblocks inside the DB.

**Comment by LlmDl** [comment]
But are you testing new outposts? I need someone to test teh following:

1. Claim an outpost.
2. Claim another outpost.
3. Travel to both of them.
4. Unclaim one outpost.
5. /ta reload.
6. Test visiting the unclaimed outpost.
7. Unclaim 2nd outpost.
8. /ta reload.
9. Test visiting both outposts.

**Comment by Portanna** [comment]
I'm currently testing the outpost thing tonight, in the meantime a player named "for" entered the server and when he gets invited to a town, he gets automatically accepted in the town but he does not appear in the town database, but just the command which listes the resident of a town, and inside the /t townname it appears several times like this:   Residents [18]: Name1, Name2, for, for, for, for, for, for, for

**Comment by LlmDl** [comment]
Please file a separate issue ticket for that For resident issue.

**Comment by Portanna** [comment]
The outpost bug seems to be fixed, I'm still testing some particular cases though. I'll let you know if there are any problems with outposts in the next days.

**Comment by LlmDl** [comment]
Thank you, it's this bug in particular that we want closed before we can do a Release Build for 1.11, 1.12 and potentially earlier builds of MC too.

**Comment by LlmDl** [comment]
Fairly confident this is fixed as of the released dev build: 0.91.4.14
